price-server/src/price_mixer.rs

```diff
@@ -0,0 +1,134 @@
+use async_trait::async_trait;
+use rust_decimal::Decimal;
+use thiserror::Error;
+
+use crate::currency::VolumePicker;
+use shared::time::*;
+use std::collections::HashMap;
+
+use super::currency::*;
+
+pub trait SidePicker {
+    fn buy_usd<'a>(&'a self) -> Box<dyn VolumePicker + 'a>;
+    fn sell_usd<'a>(&'a self) -> Box<dyn VolumePicker + 'a>;
+    fn mid_price_of_one_sat(&self) -> UsdCents;
+}
+
+#[async_trait]
+pub trait PriceProvider {
+    async fn latest(&self) -> Result<Box<dyn SidePicker>, ExchangePriceCacheError>;
+}
+
+pub struct PriceMixer {
+    providers: HashMap<String, (Box<dyn PriceProvider + Sync + Send>, Decimal)>,
+}
+
+impl PriceMixer {
+    pub fn new(
+        providers: HashMap<String, (Box<dyn PriceProvider + Sync + Send>, Decimal)>,
+    ) -> Self {
+        Self { providers }
+    }
+
+    pub async fn apply(
+        &self,
+        f: impl Fn(&Box<dyn SidePicker>) -> Decimal,
+    ) -> Result<Decimal, ExchangePriceCacheError> {
+        let mut total = Decimal::ZERO;
+        let mut total_weights = Decimal::ZERO;
+        let mut prev_error: Option<ExchangePriceCacheError> = None;
+        for (provider, weight) in self.providers.values() {
+            let side_picker = match provider.latest().await {
+                Ok(side_picker) => side_picker,
+                Err(err) => {
+                    prev_error = Some(err);
+                    continue;
+                }
+            };
+            total_weights += weight;
+            total += f(&side_picker) * weight;
+        }
+
+        if let Some(prev_error) = prev_error {
+            Err(prev_error)
```

Comment:
If you return an error here there is no point in having a `continue` at line 45. You may as well early return. Actually we want to return `Ok` even on partial error.

Reply:
I saved the previous error to make the first integration test from the price-server work: This test checks if the error NoPriceAvailabe gets propagated, if there is no tick available. 
Your comment makes sense. I just wanted to point out, that this will change existing behavior

Reply:
If none of the exchanges are available we should return an error.